### PR TITLE
First draft of the coverage of the Interbit permission model.

### PR DESCRIPTION
**Pull request by btlewan**
This definitely needs technical review.

It also needs:

- several illustrations to help cement the concepts for the reader

- cross reference links to the associated implementation discussions

- infrastructure support for Asciidoctor format
  GitBook processes Asciidoctor natively, but this involves satisfying
  Ruby dependencies

### CODE REVIEW

File: docs/gitbook/architecture/permission_model.adoc [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+Permissions are the specific actions that your application supports
+that can be assigned to roles.
+
+For example, in a voting application, the `add_vote` action which
+records a single vote on a blockchain could be assigned to the
+`can_vote` role, permitting any user with the `can_vote` role to
+dispatch the `add_vote` action.
+
+Each Interbit application includes a special permission called `*` (the
```
Comment (nicholeous) [inline docs/gitbook/architecture/permission_model.adoc]: The * is also a wildcard and can also be used to add bulk actions to a role. For example, if you had three actions:
* ADMIN_ADD_USER
* USER_DO_A_THING
* USER_DO_ANOTHER_THING

you could add a role with USER_* permissions that did all the user things.

File: docs/gitbook/architecture/permission_model.adoc [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+[[chain_joins]]
+== Chain Joins
+
+Chain joins are an Interbit-specific feature. They permit the
+information in one blockchain to be connected with information in
+another blockchain. There are two kinds of chain joins: read joins and
+write joins.
+
+A read join means that one blockchain, the _sender_, has access to a
```
Comment (nicholeous) [inline docs/gitbook/architecture/permission_model.adoc]: In a read join we refer to the sender as the provider and receiver as the consumer (analogous to the provider consumer threading model)

File: docs/gitbook/architecture/permission_model.adoc [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+copy of the state -- the collection of data stored on the blockchain --
+of another blockchain, the _recipient_, at a specific point in time. Any
+updates to the recipient's state is not reflected in the sender's
+blockchain.
+
+A write join means that, in addition to the capabilities of a read join,
```
Comment (nicholeous) [inline docs/gitbook/architecture/permission_model.adoc]: A write join means that the sender can dispatch actions into the receiver. (This is accomplished by a read join that contains a queue of actions under the hood)

File: docs/gitbook/architecture/permission_model.adoc [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+[[chain_joins]]
+== Chain Joins
+
+Chain joins are an Interbit-specific feature. They permit the
+information in one blockchain to be connected with information in
+another blockchain. There are two kinds of chain joins: read joins and
+write joins.
+
+A read join means that one blockchain, the _provider_, has access to a
+copy of the state -- the collection of data stored on the blockchain --
```
Comment (mgbarber) [inline docs/gitbook/architecture/permission_model.adoc]: We should be clear that this need not be _all_ of state. 

This has been a common misunderstanding.

File: docs/gitbook/architecture/permission_model.adoc [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+  that the user has access.
+
+- The blockchain's smart contract then compares the payload to the hash
+  to verify that the payload has not been modified. This makes actions
+  _tamper resistant_.
+
+- The blockchain's smart contract then evaluates the ACL to determine
+  whether the user has the permission to perform the action.
+
+If all of these steps complete successfully, the smart contract performs
```
Comment (mgbarber) [inline docs/gitbook/architecture/permission_model.adoc]: Is 'perform' or 'process' a better term? It feels like 'perform' is being used in two slightly different ways here.

File: docs/gitbook/architecture/permission_model.adoc [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+image:img/read_join.svg["An Interbit read join between a provider
+blockchain and a consumer blockchain", 478, 111, role="center"]
+
+A read join means that one blockchain, the _provider_, has access to a
+copy of a portion of the state -- the collection of data stored on the
+blockchain -- of another blockchain, the _consumer_, at a specific point
+in time. Any updates to the consumer's state is not reflected in the
+provider's blockchain.
+
```
Comment (mgbarber) [inline docs/gitbook/architecture/permission_model.adoc]: 'Provider' and 'consumer' are backwards in the first sentence.

**Comment by brentbenton** [comment]
"... (which is a key part of how chain joins work)."

On this topic I'd avoid using the word "key" other than for "public key", "private key".

**Pull request merged**